plugins/org.python.pydev.core/src/org/python/pydev/core/docutils/ReversePyDocIterator.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2020 by Brainwy Software Ltda
+ */
+package org.python.pydev.core.docutils;
+
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.IDocument;
+import org.python.pydev.core.IPythonPartitions;
+import org.python.pydev.core.partition.PyPartitionScanner;
+import org.python.pydev.shared_core.partitioner.FastPartitioner;
+import org.python.pydev.shared_core.string.FastStringBuffer;
+
+public class ReversePyDocIterator implements IPyDocIterator {
+
+    private int offset;
+    private IDocument doc;
+
+    private int lastReturned = -1;
+    private FastPartitioner fastPartitioner;
+
+    public ReversePyDocIterator(IDocument doc) throws BadLocationException {
+        this(doc, doc.getLineOfOffset(doc.getLength() - 1));
+    }
+
+    public ReversePyDocIterator(IDocument doc, int startingLine) throws BadLocationException {
+        this.doc = doc;
+        this.offset = doc.getLineOffset(startingLine) + doc.getLineLength(startingLine) - 1;
+        this.fastPartitioner = (FastPartitioner) PyPartitionScanner.checkPartitionScanner(doc);
+    }
+
+    @Override
+    public boolean hasNext() {
+        return offset >= 0;
+    }
+
+    @Override
+    public int getLastReturnedLine() {
+        if (offset < doc.getLength() - 1) {
+            try {
+                lastReturned = doc.getLineOfOffset(offset + 1);
+            } catch (BadLocationException e) {
+                //ignore (keep the last one)
+            }
+        }
+        return lastReturned;
+    }
+
+    /**
+     * @return the next line in the document
+     */
+    @Override
+    public String next() {
+        FastStringBuffer buf = new FastStringBuffer();
+        FastStringBuffer convertedLiteralsBuf = new FastStringBuffer();
+        boolean gotContents = false;
+        boolean gotLiterals = false;
+        try {
+            char ch = doc.getChar(offset);
+            //handle the \r, \n or \r\n
+            if (ch == '\n' || ch == '\r') {
+                offset--;
+                ch = doc.getChar(offset);
+                if (offset >= 0 && ch == '\r') {
+                    offset--;
+                    ch = doc.getChar(offset);
+                }
+            }
+
+            while (offset >= 0) {
+                ch = doc.getChar(offset);
+                if (ch == '\n') {
+                    break;
+                }
+
+                if (fastPartitioner.getContentType(offset) != IPythonPartitions.PY_DEFAULT) {
+                    convertedLiteralsBuf.insert(0, ' ');
+                    gotLiterals = true;
+                } else {
+                    buf.insert(0, ch);
```

Comment:
The contents should be appended to the end and before returning reverse it (adding to the end is cheap... to add things to the start all the existing contents must be shifted).